src/EventStore.Core/TransactionLog/Chunks/TFChunk/TFChunk.cs

```diff
@@ -1066,8 +1105,36 @@ public enum ChunkVersions : byte {
 
 		private void ReturnReaderWorkItem(ReaderWorkItem item) {
 			if (item.IsMemory) {
+				// we avoid taking the _cachedDataLock here every time because we would be
+				// contending with other reader threads also returning readerworkitems.
+				//
+				// instead we check _cacheStatus to give us a hint about whether we need to lock.
+				// but this read of _cacheStatus is not inside the lock, so it can be wrong
+				// because we read a stale value, or it changed immediately after
+				// we read it.
+				//
+				// if the check is wrong it can result in one of two outcomes:
+				// 1. we call TryDestructMemStreams when we shouldn't.
+				//    we protect against this by checking the condition again inside the lock
+				// 2. we don't call TryDestructMemStreams when we should.
+				//    the memory barrier, corresponding to the barrier in UnCacheFromMemory, protects
+				//    against this by guaranteeing ordering as emphasied in _italics_ below.
+				//    the case protected against is that another thread might call UnCacheFromMemory just as
+				//    we are returning an item here and neither thread tidys up the item we are returning.
+				//    but consider: UnCacheFromMemory sets the state to Uncaching _and then_ tidys up the
+				//    items in the pool. if it does not tidy up our item, it is because our item wasn't
+				//    in the pool. which means we hadn't enqueued it yet, which means that we will
+				//    enqueue it and then _after that_ read _cacheStatus and find it is Uncaching,
+				//    which means we will tidy it up here.
```

Comment:
do we also need to argue about the case where we don't tidy it up the memstream on this thread? :

```
if we don't tidy it here, it  means that _cacheStatus was not yet CacheStatus.Uncaching (it must have been CacheStatus.Cached). we can be sure that UnCacheFromMemory() will tidy us up as we've already crossed our barrier and thus UnCacheFromMemory will see our item in the queue and tidy it up.
```

Reply:
i don't think we need to explain about it because its the contrapositive

we explained that `!tidied_there => tidied_here`
therefore we get `!tidied_here => tidied_there` (the contrapositive) for free

and its already quite a long comment 😅

Reply:
nice :smile:  i hadn't realized that